core-rust/state-manager/src/store/traits.rs

```diff
@@ -466,3 +478,55 @@ pub mod extensions {
         ) -> Box<dyn Iterator<Item = StateVersion> + '_>;
     }
 }
+
+pub struct TransactionAndProofIterator<'a> {
+    committed_transaction_bundle:
+        Peekable<Box<dyn Iterator<Item = CommittedTransactionBundle> + 'a>>,
+    ledger_proof: Peekable<Box<dyn Iterator<Item = LedgerProof> + 'a>>,
+}
+
+impl<'a> TransactionAndProofIterator<'a> {
+    pub fn new(
+        committed_transaction_bundle: Peekable<
+            Box<dyn Iterator<Item = CommittedTransactionBundle> + 'a>,
+        >,
+        ledger_proof: Peekable<Box<dyn Iterator<Item = LedgerProof> + 'a>>,
```

Comment:
(minor) I would think that "we need to use `Peekable`" is an impl detail of the `TransactionAndProofIterator`; and the constructor could simply accept 2 iterators to be combined

Reply:
I think `Peekable` is the appropriate trait, since that is what we need/use. The fact that there is a default implementation for all `Iterator`s it's just a convenience. In the end you can implement your own `Peekable` with different/more performant (?) implementation.

Reply:
> I think Peekable is the appropriate trait

But it's a concrete impl (struct).

What's a contract of an "iterator combiner"?
I'd say it is:

"Give me 2 iterators, and I will return you some iterator which combines them [according to whatever business rules]. Ahh, and regarding my implementation details, maybe I'll use `Peekable` from standard lib, or maybe I'll do my own magic - I decide".

With this interpretation, my ideal constructor would look like:
`fn new(left: impl Iterator<A>, right: Iterator<B>) -> Self`

(please note no `Peekable` and no `Box`)

But I don't feel super strongly.

Reply:
Removal of `dyn` or `Box` would have been enough of a reason :D, but unfortunately that is not the case. While you can have `param: impl Iterator<Item = A>` you can't have `Peekable<impl Iterator<Item = A>>` - which is what you get by calling `.peekable()`.

I will make the change however. :)